---
title: "DEA_script.R"
output: html_document
date: "2025-05-16"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

# Count normalization


installing packages
```{r, eval=FALSE, message=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")


BiocManager::install("DESeq2",force = TRUE)
BiocManager::install("edgeR",force = TRUE)

```


load libraries
```{r, message=FALSE}
library(edgeR)
library(DESeq2)

```


set working directory
```{r}
#setwd("Z:/RNA_seq_DEA")

```


Here we read the data and change row names
```{r}
Coinfection.targets<-read.delim("./data/fileDesc.txt")
rownames(Coinfection.targets)<-c("SlH1","SlH2","SlH3","Ctr1","Ctr2","Ctr3")

```


read file
```{r}
Coinfection.orig <- readDGE(Coinfection.targets, header=F)

```


check dimesions
```{r}
dim(Coinfection.orig)

```


rows of data
```{r}
head(Coinfection.orig)

```


extract counts dataframe
```{r}
Coinfection.rawCount <- Coinfection.orig$count
dim(Coinfection.rawCount)

```


```{r}
head(Coinfection.rawCount)

```


## Build meta data
define sampletype
```{r}
sampletype <- factor(c(rep("SlH",3), rep("Ctr", 3)))

```


build meta data frame
```{r}
meta <- data.frame(sampletype, row.names = colnames(Coinfection.orig$count))

```


check column names
```{r}
colnames(Coinfection.orig$count)

```


rownames of meta data
```{r}
rownames(meta)

```


check that they match
```{r}
all(colnames(Coinfection.orig$count) %in% rownames(meta))

```


## Create DESeqDataset object
load the package DESeq2 Create DESEq2 dataset dds
```{r,  message=FALSE}
library(DESeq2)
dds <- DESeqDataSetFromMatrix(Coinfection.orig, colData = meta, design = ~ sampletype)

```


view dds
```{r}
head(counts(dds))

```


Generate size factors to perform median ratios
```{r}
dds <- estimateSizeFactors(dds)
sizeFactors(dds)

```


Generate normalized counts and save them to the results folder. Remember to create a results folder. 
```{r}
normalized_counts <- counts(dds, normalized=TRUE)
write.csv(normalized_counts, file="./results/coinfection_normalized_counts_DESeq2.csv")

```



# Sample-level quality Control

## Principal Component Analysis (PCA)


Transform counts for visualization
```{r}
rld <- rlog(dds, blind=TRUE)

```


plot PCA
```{r}
plotPCA(rld, intgroup="sampletype")

```


The PCA plot shows the first two principal components after variance stabilizing transformation. The plot indicates the varaiance explained by each component. In this case, the PC1 explains 40% of the variance and PC2 explains 21%). On the plot we can see that one of the control and test sample cluster right next to each other, indicating similarity between those. 


save the picture
```{r}
pdf("./results/PlotPCA_dds.pdf")
plotPCA(rld, intgroup="sampletype")
dev.off()

```


## Hierarchical Clustering Heatmap


extract rlog matrix
```{r}
rld_mat <- assay(rld)

```


compute pairwise correlation for samples
```{r}
rld_cor <- cor(rld_mat) 
head(rld_cor)

```


Here we can see the relatedness of the different samples to each other, which can be visualized more clearly using a heatmap.


head meta data for later use
```{r}
head(meta)

```


install pheatmap 
```{r, eval = FALSE}
install.packages("pheatmap")

```


load package and make a heatmap
```{r,  message=FALSE}
library(pheatmap)
pheatmap(rld_cor, annotation = meta)

```

Change sampletype color using the hex code from rbg color picker.Change coulours and add cellheight and width and save as pdf. Had to specify cells as the format changed when saved.
```{r}
pdf("./results/PlotHeatmap_dds.pdf")
heat.colors <- RColorBrewer::brewer.pal(6, "PuRd")
thing_color = list(sampletype = c(Ctr = "#eb4680", SlH = "#f5b5cb"))
pheatmap(rld_cor, annotation = meta, color = heat.colors,  border_color=NA, fontsize = 10, 
        fontsize_row = 10, height=20, cellwidth = 40, cellheight = 40, annotation_colors = thing_color)

```


Also saving it as a png as the pdf got corrupted for some reason.
```{r}
png("./results/PlotHeatmap_dds.png")
heat.colors <- RColorBrewer::brewer.pal(6, "PuRd")
thing_color = list(sampletype = c(Ctr = "#eb4680", SlH = "#f5b5cb"))
pheatmap(rld_cor, annotation = meta, color = heat.colors,  border_color=NA, fontsize = 10, 
        fontsize_row = 10, height=20, cellwidth = 40, cellheight = 40, annotation_colors = thing_color)

```


The heatmap shows the Euclidean distances between the samples after variance stabilizing transformation. The higher value on the heatmap scale (red on the first one, dark pink on the saved ones) indicates closer relation between the samples. SIH1 and SIH3 seem to be the closest related samples to each other out of all the samples. The others are not as closely related but there is relatedness seen between Ctr 1 and 2, as well as some relation between Ctr3 and SlH2, but not as close. SlH3 and Ctr1 have closer relation than Ctr3 and SlH2.  


# Differential expression analysis (DEA) using EdgeR

## Read the individual file by a meta file called fileDesc.txt
load EdgeR package
```{r,  message=FALSE}
library(edgeR)
options(digits=3)

```


Check where the data files are -- check the data sets
```{r}
infection.targets<-read.delim("./data/fileDesc.txt")

```


Check coinfection targets
```{r}
infection.targets

```


change the raw names 
```{r}
rownames(infection.targets)<-c("SlHa1","SlHa2","SlHa3","Ctr1","Ctr2","Ctr3")
infection.targets
```


Read and merge gene expression. Could have named it coinfection as that is what I am really investigating but the examplke is for Ha = infection. 
```{r}
infection <- readDGE(infection.targets, header=F)

```


Check the dimension of DGElist R object
```{r}
dim(infection)

```


Using $counts after infection gives the number in each sample. Without specifying counts it gives 2 tables. 
```{r}
head(infection$counts)

```


## Raw count distribution
First get the raw map before starting filtering
```{r}
infection.rawCount <- infection$count
head(infection.rawCount)

```


Install and load ggplot2
```{r, eval=FALSE}
install.packages("ggplot2")

```


```{r,  message=FALSE}
library(ggplot2)

```


Plot a histogram to see RNA seq distribution.
```{r}
ggplot(infection.rawCount) +
  geom_histogram(aes(x = SlHa1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")

```


Export it as a png file
```{r}
png("./results/count distribution.png", res=300, height=1800, width=1800)
ggplot(infection.rawCount) +
  geom_histogram(aes(x = SlHa1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
dev.off()
```


Make a more zoomed in plot to the area where the most genes are
```{r}
ggplot(infection.rawCount) +
  geom_histogram(aes(x = SlHa1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes") +
  xlim(0, 2000) + ylim(0, 6000)

```


## Count normalization with TMM in EdgeR
Export the raw count into the results folder previously created. 
```{r}
write.csv(infection.rawCount, file="./results/infection.rawCounts.csv")

```


Get counts per million TMM normalised BEFORE filtering
```{r}
infection.normCPM <- cpm(calcNormFactors(infection))
dim(infection.normCPM)

```


```{r}
head(infection.normCPM)

```


Save the normalised results
```{r}
write.csv(infection.normCPM, file="./results/infection.normCPM.csv")

```


## Filter counts per million (CPM) out and get smaller size of libraries
Keep genes that are expressed at least 1 CPM in at least 3 libraries, normally it is the number of biological replicates of smaller group
```{r}
infection.filtered <- rowSums(cpm(infection)>1) >=3
table(infection.filtered)
```


Libraries size of data BEFORE filtering
```{r}
infection$samples$lib.size

```


Make an infection file with the filtered data
```{r}
Infection <- infection[infection.filtered,]

```


Library size after filtering
```{r}
colSums(Infection$counts)

```


```{r}
dim(Infection)

```


Update library size 
```{r}
Infection$samples$lib.size <- colSums(Infection$counts)
Infection$samples
```


Here as comparison before filtering --> not a huge change but still changes
```{r}
infection$samples

```


## Count normalization after filter
Performed normalisation with TMM method
```{r}
Infection = calcNormFactors(Infection)

```


The libraries after normalisation
```{r}
Infection$samples

```


Get the counts per million (TMM normalised) after filtering
```{r}
Infection.filtered.normCPM <-cpm(calcNormFactors(Infection))

```


Export TMM normalized count table AFTER filtering
```{r}
write.csv(Infection.filtered.normCPM, file="./results/Infection.filtered.normCPM.csv")

```


## Experimental design
Treatment factor
```{r}
group<-factor(c('SlHa','SlHa','SlHa',"Ctr","Ctr","Ctr"))

```


Describe the experimental design,one factor with intercept, here Ctr is the intercept
```{r}
Infection.design <- model.matrix(~group)   
rownames(Infection.design)<-colnames(Infection$counts)
Infection.design

```


## Sample-level quality control by multidimensional scaling (MDS)
Make an MDS plot to visualise outliers --> shows relative similarities between the samples
```{r}
plotMDS(Infection, main="MDS plot of RNA-Seq", labels=colnames(Infection$counts))

```


Save as a png
```{r}
png("./results/plotMDS.Infection.png", res=300, height=1800, width=1800)
plotMDS(Infection, main="MDS plot of Infection RNA-Seq", labels=colnames(Infection$counts))
dev.off()

```


The MDS plot shows distances corresponding to the logFC between each of the samples.  The leading LogFC is the base 2 logarithm of fold change and is the average of the largest logFC between each pair of samples. The plot indicates that SlHa3 and Ctr2 are the closest related two samples as they are closest together on the plot. SlHa1 is also pretty closely related to SlHa3.


## Estimating the common dispersion, trended dispersion, tagwwise dispersion
Dispersion means biological coeffient of variation (BCV) squared.E.g. if genes expression typically differs from replicate to replicate by 20% its BCV is 0.2, and its dispersion is 0.04. Estimating the common dispersion
```{r}
Infection <- estimateGLMCommonDisp(Infection, Infection.design)

```


Estimating the trended dispersion
```{r}
Infection <- estimateGLMTrendedDisp(Infection, Infection.design)

```


Estimating the tagwwise dispersion
```{r}
Infection <- estimateGLMTagwiseDisp(Infection, Infection.design)

```


Creating a visual representation of the mean-variance relationship and BCV-mean logCPM relationship
```{r}
plotMeanVar(Infection, show.tagwise.vars=T,NBline=T)
plotBCV(Infection)

```


dispersion is the coefficient of biological variation (BCV). The common BCV is on the high side, considering that this is a designed experiment using genetically identical plants. The trended dispersion shows a decreasing trend with expression level. At low logCPM, the dispersions are very large.


## Fit DGEList and design matrix to genewise GLM

```{r}
Infection.fit <- glmFit(Infection, Infection.design)
colnames(Infection.fit)
```


##Pairwise comparison


Likelihood test for genes respond differently between different conditions, table of top differentially expressed tags, n specify n tags to display
```{r}
lrt.SlHa_vs_Ctr <- glmLRT(Infection.fit, coef=2)  # to compare SlHa vs Ctr (SlHa_vs_Ctr)

```


```{r}
t1<-topTags(lrt.SlHa_vs_Ctr, n=nrow(Infection))
head(t1$table)
```


## Extract number of differentially expressed (DE) genes
```{r}
summary(decideTests(lrt.SlHa_vs_Ctr, adjust.method="BH", p.value=0.05))

```


The number of UP-regulated genes (can change logFC to be more strict)
```{r}
nrow(subset(topTags(lrt.SlHa_vs_Ctr, n=586)$table,  logFC > 0))

```

lrt is the likelihood ratio test. Here we make a dataset for it that is stored where we can see the PITA that are upregulated
```{r}
lrt.SlHa_vs_Ctr_UP <- subset(topTags(lrt.SlHa_vs_Ctr, n=586)$table, logFC > 0)

```


The number of  DW-regulated genes (can change logFC to be more strict)
```{r}
nrow(subset(topTags(lrt.SlHa_vs_Ctr, n=586)$table,  logFC < 0))

```


Make dataset of the DW regulated 
```{r}
lrt.SlHa_vs_Ctr_DW <- subset(topTags(lrt.SlHa_vs_Ctr, n=586)$table, logFC < 0)

```


## Differentially expressed transcriptsâ€™ tag
```{r}
DEtags.lrt.SlHa_vs_Ctr <- rownames(Infection)[as.logical(decideTests(lrt.SlHa_vs_Ctr, adjust.method="BH", p.value=0.05))]

```


Export list of UP-regulated and DW-regulated transcripts
```{r}
write.csv(lrt.SlHa_vs_Ctr_UP, file="./results/lrt.SlHa_vs_Ctr_UP.csv")
write.csv(lrt.SlHa_vs_Ctr_DW, file="./results/lrt.SlHa_vs_Ctr_DW.csv")

```


Genes label as grey color
```{r}
Infection.colHavsCtr = rep('grey55', nrow(Infection))

```


Assign color for DE trasncripts
```{r}
Infection.colHavsCtr[lrt.SlHa_vs_Ctr$table$PValue < 0.05 & lrt.SlHa_vs_Ctr$table$logFC >0 ] <- "red"
Infection.colHavsCtr[lrt.SlHa_vs_Ctr$table$PValue < 0.05 & lrt.SlHa_vs_Ctr$table$logFC <0 ] <- "blue"
```


Plot Smear plot with adjustment on Y-axis label
```{r}
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.SlHa_vs_Ctr, de.tags=DEtags.lrt.SlHa_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="SlHa infection compared to Control", pch=19, cex=0.4, smearWidth=0.5, panel.first=grid(), smooth.scatter=FALSE, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")


```


adjustment on Y-axis label and colour of DE tags changed
```{r}
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.SlHa_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="a infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")


```


Plot Smear plot with adjustment on Y-axis label and colour of DE tags changed and export as png

```{r}
png("./results/plotSmear.InfectionRNAseq.png", res=300, height=1800, width=1800)
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.SlHa_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="SlnHa infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")
       
```

The plots show the difference in gene expression when coinfected compared to control. So here, the upregulated genes are shown in red while the downregulated genes are shown in blue. The genes that remain grey in the middle are unchanged.

Done.